qubesadmin/tools/qvm_ls.py

```diff
@@ -710,16 +786,109 @@ def main(args=None, app=None):
             vm for vm in args.app.domains if vm.name not in args.exclude
         ]
 
+    if args.klass:
+        # filter only VMs to specific class(es)
+        domains = [d for d in domains if d.klass in args.klass]
+
+    if args.label:
+        # filter only VMs with specific label(s)
+        domains_labeled = []
+        spinner.show('Filtering based on labels...')
+        for dom in domains:
+            if dom.label.name in args.label:
+                domains_labeled.append(dom)
+            spinner.update()
+        domains = domains_labeled
+        spinner.hide()
+
     if args.tags:
         # filter only VMs having at least one of the specified tags
         domains = [dom for dom in domains
                    if set(dom.tags).intersection(set(args.tags))]
 
+    if args.no_tags:
+        # exclude VMs having at least one of the specified tags
+        domains = [dom for dom in domains
+                   if not set(dom.tags).intersection(set(args.no_tags))]
+
+    if args.based_on:
+        # Filter only VMs based on specific TemplateVM
+        child_domains = []
```

Comment:
Please use a name that reference the it is child domain of a template. The function `tree_append_child()` uses `childs` variable for network children. Use something similar to `domains_child_of_template`. Maybe child is redundant in this name.

Also you use `child_domains` here and `domains_connecting` later. Use a consistent naming scheme, such as beginning always with `domains_` and referencing the connection type later.